Update performance of global styles code

oandregal opened the pull request:
This PR improves the performance of global styles by not looking up for resources we don't need.

What this means in practice:

- Styles: we no longer enqueue the global styles stylesheet if a theme doesn't have theme.json. We used to do that to make the presets available (both CSS Custom Properties & classes) for all themes automatically. People have reported this is not ideal and may interfere with their themes' CSS.

- Settings: whether or not the theme has a theme.json, we need to enqueue the settings to the editors to control design tools visibility, etc. This is now bounded to not require user data lookup if it has no support.

### Test

- Themes that support FSE & Global Styles work as expected ― theme example: TT1-blocks
  - site-editor: changes in the user styles are reflected in the front-end
  - post-editor: still works as expected (presets are present, etc)

- Themes that don't support FSE and/or Global Styles work as expected ― theme example: TwentyTwentyOne, TwentyTwenty.
  - site-editor: is not loaded for them
  - post-editor: still works as expected (presets are present, etc)

oandregal committed:
Early return for public methods if theme does not have support

oandregal committed:
Early return as we do not register the user CPT if no theme.json

oandregal committed:
Do not bind hooks if no theme.json present

oandregal committed:
Settings: do not check for theme/user data if no theme.json

jorgefilipecosta commented on lib/class-wp-theme-json-resolver.php:
```diff
@@ -270,6 +274,10 @@ private static function get_user_origin() {
 	 * @return WP_Theme_JSON
 	 */
 	public function get_origin( $theme_support_data = array(), $origin = 'user', $merged = true ) {
+		if ( ! gutenberg_experimental_global_styles_has_theme_json_support() ) {
```
I'm not sure if should be this function verifying if theme.json is supported or not. If this function is called with an origin user  I guess it should return the origin user. The caller of this function should probably be the one making the checks?

jorgefilipecosta commented on lib/class-wp-theme-json-resolver.php:
```diff
@@ -180,6 +180,10 @@ private function get_theme_origin( $theme_support_data = array() ) {
 	 * @return array Custom Post Type for the user's origin config.
 	 */
 	private static function get_user_data_from_custom_post_type( $should_create_cpt = false, $post_status_filter = array( 'publish' ) ) {
+		if ( ! gutenberg_experimental_global_styles_has_theme_json_support() ) {
```
The same comment as in get_origin, it seems this function should always return the user custom_post_type and the checks should not be here.

I guess the checks should be in gutenberg_experimental_global_styles_enqueue_assets and gutenberg_experimental_global_styles_settings?

jorgefilipecosta commented on lib/global-styles.php:
```diff
@@ -235,6 +243,9 @@ function_exists( 'gutenberg_is_edit_site_page' ) &&
 }
 
-add_action( 'init', array( 'WP_Theme_JSON_Resolver', 'register_user_custom_post_type' ) );
+if ( gutenberg_experimental_global_styles_has_theme_json_support() ) {
```
I think the function gutenberg_experimental_global_styles_has_theme_json_support  must be called inside actions. Right now with this change, I'm getting some notices:
```

  | <br />
-- | --
  | <b>Notice</b>:  Use of undefined constant STYLESHEETPATH - assumed 'STYLESHEETPATH' in <b>/Users/pc/dev/core/wordpress-develop/build/wp-includes/template.php</b> on line <b>663</b><br />
  | <br />
  | <b>Notice</b>:  Use of undefined constant TEMPLATEPATH - assumed 'TEMPLATEPATH' in <b>/Users/pc/dev/core/wordpress-develop/build/wp-includes/template.php</b> on line <b>666</b><br />
  | <br />
  | <b>Warning</b>:  Cannot modify header information - headers already sent by (output started at /Users/pc/dev/core/wordpress-develop/build/wp-includes/template.php:663) in <b>/Users/pc/dev/core/wordpress-develop/build/wp-includes/functions.php</b> on line <b>6360</b><br />
  | <br />
  | <b>Warning</b>:  Cannot modify header information - headers already sent by (output started at /Users/pc/dev/core/wordpress-develop/build/wp-includes/template.php:663) in <b>/Users/pc/dev/core/wordpress-develop/build/wp-admin/includes/misc.php</b> on line <b>1310</b><br />
  | <br />
  | <b>Warning</b>:  Cannot modify header information - headers already sent by (output started at /Users/pc/dev/core/wordpress-develop/build/wp-includes/template.php:663) in <b>/Users/pc/dev/core/wordpress-develop/build/wp-admin/admin-header.php</b> on line <b>9</b><br />

```

I think it happens because some constants required are not initialized during this code execution.

jorgefilipecosta commented:
I think we are causing a regression, in themes with a theme.json file e.g: Twenty Twenty-One Blocks we are outputting the expected styles.

oandregal committed:
Use kses to filter out invalid user content

oandregal committed:
Move checks to proper place

oandregal committed:
Move checks to consumers of the API

oandregal committed:
Revert changes, improve diff readibility

oandregal committed:
Improve code

oandregal committed:
fixup variable name

oandregal committed:
Add phpdocs

oandregal committed:
Make linter happy

nosolosw commented:
@jorgefilipecosta your feedback was addressed.

jorgefilipecosta commented on lib/global-styles.php:
```diff
@@ -162,6 +162,10 @@ function gutenberg_experimental_global_styles_get_stylesheet( $tree, $type = 'al
  * and enqueues the resulting stylesheet.
  */
 function gutenberg_experimental_global_styles_enqueue_assets() {
+	if ( ! gutenberg_experimental_global_styles_has_theme_json_support() ) {
```
One of the advantages global styles brings is that themes don't need to create preset classes anymore. With this change I guess we remove this advantage right? I guess if theme.json is not present we don't need to make database queries and retrieve user settings but I guess we can still output preset classes?

nosolosw replied:
Yes, one of the goals of this PR is that we will no longer enqueue any CSS **if the theme doesn't have theme.json support**. This is a mechanism to make sure that 1) we don't interfere with the theme's CSS and 2) we don't load unnecessary styles. The current behavior has been reported as not ideal in core-editor by people. I still think it's valuable if the theme supports theme.json, though.

jorgefilipecosta reviewed (commented):
It seems in this branch the color palettes from the themes are not being applied and the default ones are always being used. I tested with the 2020 theme.

nosolosw commented:
> It seems in this branch the color palettes from the themes are not being applied and the default ones are always being used. I tested with the 2020 theme.

oh! good catch, thank you.

oandregal committed:
Enqueue settings for all themes

The presets are passed via settings, so whether or not the theme
supports theme.json, we need to retrieve the editor settings
and pass them back to the client.

oandregal committed:
Make linter happy

oandregal committed:
Update styles

oandregal committed:
Process border settings

nosolosw commented on lib/class-wp-theme-json.php:
```diff
@@ -337,6 +339,7 @@ public function __construct( $contexts = array() ) {
 			// Process settings subtree.
 			$this->process_key( 'settings', $context, self::SCHEMA );
 			if ( isset( $context['settings'] ) ) {
+				$this->process_key( 'border', $context['settings'], self::SCHEMA['settings'] );
```
This is unrelated and should have been added in the PR that added the border as settings, but I'm adding here.

nosolosw commented on lib/class-wp-theme-json.php:
```diff
@@ -324,8 +325,9 @@ public function __construct( $contexts = array() ) {
-				$this->process_key( 'color', $context['styles'], self::SCHEMA['styles'] );
-				$this->process_key( 'typography', $context['styles'], self::SCHEMA['styles'] );
+				$this->process_key( 'color', $context['styles'], self::SCHEMA['styles'], $should_escape_styles );
+				$this->process_key( 'spacing', $context['styles'], self::SCHEMA['styles'], $should_escape_styles );
```
This is unrelated but I've noticed that was missing, so I'm adding it here.

nosolosw commented on lib/global-styles.php:
```diff
@@ -196,45 +200,74 @@ function gutenberg_experimental_global_styles_settings( $settings ) {
+	// even for themes that don't support theme.json.
+	// An example of this is that the presets are configured
+	// from the theme support data.
+	$settings['__experimentalFeatures'] = $tree->get_settings();
```
This will still look up for the theme.json, although the theme doesn't have one. I'd like to squeeze this performance bit as well, but it was getting convoluted so I'd like to do it in a follow-up (it requires improving the `WP_Theme_JSON_Resolver` API for consumers). So, this PR improves a bit the situation but doesn't fix all the things.

nosolosw commented on lib/global-styles.php:
```diff
@@ -196,45 +200,74 @@ function gutenberg_experimental_global_styles_settings( $settings ) {
 		gutenberg_is_edit_site_page( $screen->id ) &&
-		gutenberg_experimental_global_styles_has_theme_json_support()
+		gutenberg_experimental_global_styles_has_theme_json_support() &&
+		gutenberg_is_fse_theme()
```
We have a couple of checks that are related but independent:

- whether a theme is FSE-enabled (has a `block-templates/index.html`) => this is the check we use to decide whether to load the site editor, hence we also check for this when deciding whether to lookup user styles for the site editor.
- whether a theme is global-styles enabled (has an `experimental-theme.json`) => if a theme doesn't have theme.json it doesn't make sense to offer user styles, as 1) this are merged with the precedent origins (core, theme) and 2) we don't know if the user styles will play well with the theme's CSS.

oandregal committed:
Fix rebase issues with linter

jorgefilipecosta reviewed (approved):
LGTM 👍

hoangitt commented:
Hello team, 
I'm using wordpress 5.6. I created file theme.json but not worked in my site
Structure JSON FILE:

`{
    "global": {
        "settings": {
            "custom": {
                "typo": {
                    "rootSize": 16,
                    "scale": "var(--viweb-paragraph-font-size-scale)"
                },
                "line-height": {
                    "small": 1.2,
                    "medium": 1.4,
                    "large": 1.8
                }
            },
            "spacing": {
                "customPadding": true,
                "units": ["px", "em", "rem", "vh", "vw"]
            },
            "color": {
                "palette": [
                    {
                        "slug": "primary",
                        "color": "var(--viweb-palette-primary-color)"
                    },
                    {
                        "slug": "second",
                        "color": "var(--viweb-palette-second-color)"
                    },
                    {
                        "slug": "orange",
                        "color": "var(--viweb-palette-orange-color)"
                    },
                    {
                        "slug": "blue",
                        "color": "var(--viweb-palette-blue-color)"
                    }
                ]
            },
            "typography": {
                "fontSizes": [
                    {
                        "slug": "small",
                        "size": "calc(var(--viweb-paragraph-font-size-small) / var(--viweb-paragraph-font-size-scale))"
                    },
                    {
                        "slug": "medium",
                        "size": "calc(var(--viweb-paragraph-font-size-medium) * 1px)"
                    },
                    {
                        "slug": "large",
                        "size": "calc(var(--viweb-paragraph-font-size-large) * var(--viweb-paragraph-font-size-scale))"
                    },
                    {
                        "slug": "huge",
                        "size": "calc(var(--viweb-paragraph-font-size-huge) * var(--viweb-paragraph-font-size-scale))"
                    }
                ],
                "customFontSize": true,
                "customLineHeight": true
            }
        }
    },
    "core/heading/h1": {
        "styles": {
            "typography": {
                "fontSize": "var(--viweb-heading-h1-font-size)"
            }
        }
    },
    "core/heading/h2": {
        "styles": {
            "typography": {
                "fontSize": "var(--viweb-heading-h2-font-size)"
            }
        }
    },
    "core/heading/h3": {
        "styles": {
            "typography": {
                "fontSize": "var(--viweb-heading-h3-font-size)"
            }
        }
    },
    "core/heading/h4": {
        "styles": {
            "typography": {
                "fontSize": "var(--viweb-heading-h4-font-size)"
            }
        }
    },
    "core/heading/h5": {
        "styles": {
            "typography": {
                "fontSize": "var(--viweb-heading-h5-font-size)"
            }
        }
    },
    "core/heading/h6": {
        "styles": {
            "typography": {
                "fontSize": "var(--viweb-heading-h6-font-size)"
            }
        }
    }
}`

nosolosw commented:
@hoangitt to use the `experimental-theme.json` you need to use the Gutenberg plugin, it wasn't released as part of WordPress 5.6.

hoangitt commented:
@nosolosw yep, thanks for reply
